crates/factor-sqlite/src/runtime_config/spin.rs

```diff
@@ -0,0 +1,237 @@
+    // this once, we use a `OnceCell` to store it.
+    inner: OnceCell<spin_sqlite_libsql::LibsqlClient>,
+}
+
+impl LibSqlConnection {
+    fn new(url: String, token: String) -> Self {
+        Self {
+            url,
+            token,
+            inner: OnceCell::new(),
+        }
+    }
+
+    async fn get_client(&self) -> Result<&spin_sqlite_libsql::LibsqlClient, v2::Error> {
+        self.inner
+            .get_or_try_init(|| async {
+                spin_sqlite_libsql::LibsqlClient::create(self.url.clone(), self.token.clone())
+                    .await
+                    .context("failed to create SQLite client")
+            })
+            .await
+            .map_err(|_| v2::Error::InvalidConnection)
+    }
+}
+
+#[async_trait::async_trait]
+impl Connection for LibSqlConnection {
+    async fn query(
+        &self,
+        query: &str,
+        parameters: Vec<v2::Value>,
+    ) -> Result<v2::QueryResult, v2::Error> {
+        let client = self.get_client().await?;
+        <spin_sqlite_libsql::LibsqlClient as spin_sqlite::Connection>::query(
+            client, query, parameters,
+        )
+        .await
+    }
+
+    async fn execute_batch(&self, statements: &str) -> anyhow::Result<()> {
+        let client = self.get_client().await?;
+        <spin_sqlite_libsql::LibsqlClient as spin_sqlite::Connection>::execute_batch(
+            client, statements,
+        )
+        .await
+    }
+}
+
+/// Configuration for a local SQLite database.
+#[derive(Clone, Debug, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct LocalDatabase {
+    pub path: Option<PathBuf>,
+}
+
+impl LocalDatabase {
+    /// Create a new connection pool for a local database.
+    fn pool(self, base_path: Option<&Path>) -> anyhow::Result<SimpleConnectionPool> {
+        let location = match self.path {
+            Some(path) => {
+                // TODO: `base_path` should be passed in from the runtime config
+                let path = resolve_relative_path(&path, base_path)?;
+                // Create the store's parent directory if necessary
+                // unwrapping the parent is fine, because `resolve_relative_path`` will always return a path with a parent
+                std::fs::create_dir_all(path.parent().unwrap())
+                    .context("Failed to create sqlite database directory")?;
+                spin_sqlite_inproc::InProcDatabaseLocation::Path(path)
+            }
+            None => spin_sqlite_inproc::InProcDatabaseLocation::InMemory,
+        };
+        let factory = move || {
+            let connection = spin_sqlite_inproc::InProcConnection::new(location.clone())?;
+            Ok(Arc::new(connection) as _)
+        };
+        Ok(SimpleConnectionPool::new(factory))
+    }
+}
+
+/// Resolve a relative path against an optional base path.
+fn resolve_relative_path(path: &Path, base_path: Option<&Path>) -> anyhow::Result<PathBuf> {
```

Comment:
This could probably moved to a shared place for factors as it is a replacement for `resolve_config_path` which both KV and SQLite use

Reply:
This is a good point. I think the question is _where_. It's a highly generic yet super specific piece of code... I think until it becomes obvious where this should live, duplicating isn't the worst thing in the world.